ipd/0020/README.adoc

```diff
@@ -0,0 +1,770 @@
+module that execises the various stream APIs in `genunix`.
+
+Test modules, like system libraries, should come welded to the system
+-- the source code for the test module should live in illumos-gate.
+The main exception would be a test delivered as part of an out-of-gate
+driver or for downstream distributions testing their own kernel
+functionality (though in that case it should be in their downstream
+gate).
+
+Delivering test modules is a choice left to each downstream
+distribution. That said, we must make a default decision about how to
+structure the IPS manifests in gate. First, it seems to make sense to
+at least give the ktest facility its own package, which includes only
+the means to register, list, and execute tests, but does not deliver
+any tests itself. Things get more interesting when determining how
+test modules should be delivered. The following is a table of
+potential options and their tradeoffs.
+
+|===
+|Delivery| Tradeoffs
+
+|1. All in-gate tests delivered in ktest package. Deliver all in-gate
+ test modules as part of the ktest package.
+a|* One package gives you everything.
+* No test modules delivered unless you absoltely want them.
+* Delivers test modules for modules that may not be attached and that
+have no relevance to your system .
+
+|2. Each test module is delivered with whatever package delivers the
+ module-under-test. Each package which delivers a test module has a
+ dependency on ktest facility package.
+a|* Only the necessary test
+ modules are installed.
+* Probably makes the most logical sense.
+* Given that at least one module-under-test is part of the main kernel
+  (like genunix), this effectively means ktest is always delivered.
+
+|3. Same as previous, but don't require ktest dependency.
+a|* Same benefits as above, but test execution can only happen if the
+user decides to also install ktest. Otherwise the test modules lay
+doormant on the filesystem (not loaded).
+
+|===
+
+I think we should go with option (3). We should deliver test-modules
+with their module-under-test, but only load/run them when the ktest
+facility is installed (and even then they would not be loaded until
+the user specifically requests that one or more test-modules be
+loaded).
+
+As these test-modules are misc-type modules, they should be delivered
+in the misc module directories. However, in order not to pollute the
+`misc/` directory, it makes sense to place them under a `ktest/`
+subdir.
+
+.ktest test modules home
+----
+/kernel/misc/ktest/amd64
+/usr/kernel/misc/ktest/amd64
+----
+
+=== ktest(1M)
+
+The `ktest(1M)` command controls all interactions between the user and
+ktest facility, as well as all interactions between the test modules
+and ktest facility. That is, unless done through some other means like
+`modload`, all test module loading, unloading, listing, and running
+should only occur as a direct result of executing the `ktest` command.
+
+The ktest device may only be accessed by root from the Global Zone.
```

Comment:
Is there a specific required privilege for this?

Reply:
I looked at the code and answered my own question - it's PRIV_SYS_DEVICES
Suggest updating the text to say this rather than 'root'

Reply:
Actually /usr/kernel/... is more appropriate for anything not required at boot time (and tests are certainly in that category)

Reply:
Okay this is good to know because I couldn't really find any good documentation about when any of these directories should be preferred.

Reply:
> I looked at the code and answered my own question - it's PRIV_SYS_DEVICES
> Suggest updating the text to say this rather than 'root'

Yea that's better, I'm actually already using `drv_priv(9F)` in the driver anyways.

> Actually /usr/kernel/... is more appropriate for anything not required at boot time (and tests are certainly in that category) 

I updated the text to only include the `/usr/kernel/...` path.